include/exadg/incompressible_navier_stokes/postprocessor/postprocessor.cpp

```diff
@@ -101,53 +104,267 @@ PostProcessor<dim, Number>::setup(Operator const & pde_operator)
+    sol.name        = "mean_velocity";
+    sol.dof_handler = &navier_stokes_operator->get_dof_handler_u();
+    sol.vector      = &mean_velocity;
+    this->additional_fields.push_back(sol);
+  }
+
+  // cfl
+  if(pp_data.output_data.write_cfl)
+  {
+    SolutionField<dim, Number> sol;
+    sol.type   = SolutionFieldType::cellwise;
+    sol.name   = "cfl_relative";
+    sol.vector = &cfl_vector;
+    this->additional_fields.push_back(sol);
+  }
+}
+
+template<int dim, typename Number>
+void
+PostProcessor<dim, Number>::compute_mean_velocity(VectorType &       mean_velocity,
+                                                  VectorType const & velocity)
+{
+  unsigned int const counter = time_control_mean_velocity.get_counter();
+  mean_velocity.sadd((double)counter, 1.0, velocity);
+  mean_velocity *= 1. / (double)(counter + 1);
+}
+
+template<int dim, typename Number>
+void
+PostProcessor<dim, Number>::calculate_additional_vectors(VectorType const &     velocity,
+                                                         double const           time,
+                                                         types::time_step const time_step_number)
+{
+  bool vorticity_is_up_to_date = false;
+  if(pp_data.output_data.write_vorticity == true)
+  {
+    navier_stokes_operator->compute_vorticity(vorticity, velocity);
+    vorticity_is_up_to_date = true;
+  }
+
+  if(pp_data.output_data.write_divergence == true)
+  {
+    navier_stokes_operator->compute_divergence(divergence, velocity);
+  }
+
+  if(pp_data.output_data.write_velocity_magnitude == true)
+  {
+    navier_stokes_operator->compute_velocity_magnitude(velocity_magnitude, velocity);
+  }
+
+  if(pp_data.output_data.write_vorticity_magnitude == true)
+  {
+    AssertThrow(vorticity_is_up_to_date == true,
+                dealii::ExcMessage(
+                  "Vorticity vector needs to be updated to compute its magnitude."));
+
+    navier_stokes_operator->compute_vorticity_magnitude(vorticity_magnitude, vorticity);
+  }
+
+  if(pp_data.output_data.write_streamfunction == true)
+  {
+    AssertThrow(vorticity_is_up_to_date == true,
+                dealii::ExcMessage(
+                  "Vorticity vector needs to be updated to compute its magnitude."));
+
+    navier_stokes_operator->compute_streamfunction(streamfunction, vorticity);
+  }
+
+  if(pp_data.output_data.write_q_criterion == true)
+  {
+    navier_stokes_operator->compute_q_criterion(q_criterion, velocity);
+  }
+
+  if(pp_data.output_data.mean_velocity.is_active == true)
+  {
+    // TODO: THIS IS A QUESTION IN GENERAL: SHOULD THIS ASSERT BE CHECKED INSIDE OR OUTSIDE? I TEND
+    // TO DO IT INSIDE THE FUNCTIONS SINCE THE INTERFACE WILL LOOK THE SAME AND THE USER DOES NOT
+    // HAVE TO CARE DURING IMPLEMENTATION
+    AssertThrow(Utilities::is_unsteady_timestep(time_step_number),
+                dealii::ExcMessage(
+                  "Calculating mean velocity does not make sense for unsteady problems."));
```

Comment:
I didn't do this in a consequent way, since it was already inconsistent before. I would be in favor to move those asserts to the relevant ```evaluate()``` functions, which then all need a ```bool const unsteady```. This way the interface is more general, and we don't have to think about writing an assent if a certain calculator is reused somewhere elese.

Reply:
That a good idea. But it should read "steady problems" in the message, right?

Reply:
Yes :)

Reply:
[1c3277b](https://github.com/exadg/exadg/pull/262/commits/1c3277b32be3cc0569f9ebeb89a79db1744ef87d)